### ISSUE TITLE
Might want to drop the local and OF preprocessor defines

### ISSUE THREAD

**Issue by schwehr** [madler/zlib#117]
Hi,

I just submitted a patch to minizip in GDAL that you might be interested in (aside from the C++ parts):

https://trac.osgeo.org/gdal/changeset/31508

The 2 main things are to drop the `local` and `OF` pre-processor `#defines`.  It seems like modern debuggers should all be able to handle static functions and worrying about K&R C support these days is not worth it.  If you disagree, I'd love to hear why.  I tried looking at the mailing list archives to see if this had been discussed, but they 404 at the moment.

Pardon the C++ noise in the GDAL patch.

**Comment by tbeu** [comment]
Check the (non-official) new generation [zlib](https://github.com/Dead2/zlib-ng) for a forked project which also has the aim to remove these old-style declarations.

**Comment by schwehr** [comment]
@tbeu , Thanks for the pointer

**Comment by gvollant** [comment]
What is the PRO/CONS argument about removal ?

 

OF was used to be compatible with old C compiler

 

What is problem with it?

 

Regards

Gilles Vollant

 

De : Kurt Schwehr [mailto:notifications@github.com] 
Envoyé : Tuesday, November 17, 2015 5:24 PM
À : madler/zlib
Objet : Re: [zlib] Might want to drop the local and OF preprocessor defines (#117)

 

@tbeu <https://github.com/tbeu>  , Thanks for the pointer

**Comment by schwehr** [comment]
There is nothing wrong with it from a C perspective.  The arguments for removal are along the lines of these.  And it's just a value call for you which way to go.

- Are there K&R C only compilers that still need to be supported?  I would assume (but could be wrong) that anyone in a space that is still stuck with K&R isn't going to be allowed to use a new version of zlib anyway
- Removing them makes the code easier to read for others - most C & C++ coders active today grep up with ANSI 89 C or newer (e.g. have been coding less than 25 years)
- Removing macros helps static analyzers (e.g. coverity) do better.  Less complexity for them to track / grok.
- It certainly can't hurt runtime analysis to have less C pre-processor manipulation between the output of something like AFL [American Fuzzy Lop](http://lcamtuf.coredump.cx/afl/) and the user.

While zlib is not NTP, some of the discussion by ESR in his [NTPsec is not quite a full rewrite](http://esr.ibiblio.org/?p=6881) goes into more detail.  Even if I don't necessarily agree with everything in the article and comments, I think it's a very worthwhile read.

If there really is an continued need to support folks in a K&R C world, maybe a fork for them would be a good idea.

But again, I started with "Might" in the issue title.  It would be interesting to hear why you decide one way or the other on this issue.

**Comment by madler** [comment]
Why remove `local`?

**Comment by schwehr** [comment]
What compiler/debugger are you trying to support with `local`? From a technical point of view, I see nothing actually wrong with local and I'm sure this was definitely important >10 years ago.

If you have something specific that is still in common use, then local makes sense.  It would be nice to document what the broken toolchain is so maybe someone can fix it.  Otherwise, it's C preprocessor noise to someone not familiar with the code that adds to the cognitive load of someone is debugging a large stack.

This is the kind of stuff I spend time changing with GDAL.  While many of my changes don't seem like much, together they make maintaining and supporting a large number of programs much easier.

Likely the discussion around this suggestion is more important than if you decide to do `local` -> `static` or not.

**Comment by madler** [comment]
It's not for any compiler. It's for humans reading the code. `local` conveys the intent to make the function or variable local.

**Comment by schwehr** [comment]
Sure.  My point is:  To an external coder, I have to look up what you mean by local.  C and C++ programmers are expected to know what a static function is.  e.g.

http://stackoverflow.com/questions/558122/what-is-a-static-function

If I see `static int foo(int bar) { return bar + 1;}`, then I know without looking it's local.   Using a #define like `local` is far less common.  Coming into it, finding a CPP directive implies that there is very possibly something funky going on.  I searched and found nothing, but I had to search.  That included looking at the configuration system.  Working across a lot of code bases, it's split between "surprise, this macro thing is dealing with a real compiler / tool chain issue" and (like here) a convenience thing.  I've even seen things commented to imply it's no big deal, but later found that it was a for handling an actual issue.

I don't think I've got any more to contribute to the discussion.

Make your call either way and close the issue when you've decided to leave it or change it. 

Thank you all for taking the time to respond and for your work on zlib!
